### PR TITLE
Bug 1926893: only override deployment resources when explicitly defined in subscription config

### LINKED ISSUE #1912: If resource override is not defined in a subscription then the existing resource definition in the CSV is deleted
## Bug Report

**What did you do?**
Created the following subscription where there is no override on the resources:
```
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: consul
  namespace: consul-stable
spec:
  channel: stable
  installPlanApproval: Manual
  name: consul
  source: applications
  sourceNamespace: olm
  startingCSV: consul.v0.0.10
```
This is the CSV that has resources section defined:
```
apiVersion: operators.coreos.com/v1alpha1
kind: ClusterServiceVersion
metadata:
  annotations:
    capabilities: Basic Install
  name: consul.v0.0.10
  namespace: placeholder
spec:
  apiservicedefinitions: {}
  replaces: consul.v0.0.9
  customresourcedefinitions:
    owned:
    - kind: Consul
      name: consuls.application.com
      version: v1alpha2
      displayName: Consul operator
      description: Consul operator      
  description: Placeholder description
  displayName: Consul
  install:    
    spec:
      deployments:
      - name: consul-operator
        spec:
          replicas: 1
          selector:
            matchLabels:
              name: consul-operator
          strategy: {}
          template:
            metadata:
              labels:
                name: consul-operator
            spec:
              containers:
              - command:
                - consul-operator
                env:
                - name: WATCH_NAMESPACE
                  valueFrom:
                    fieldRef:
                      fieldPath: metadata.annotations['olm.targetNamespaces']
                image: consul-operator:1.19
                imagePullPolicy: IfNotPresent
                name: consul-operator
                resources:
                  limits:
                    memory: 384Mi
                    cpu: 250m
              serviceAccountName: consul-operator
              imagePullSecrets:
                - name: pull-secret
      permissions:
      - rules:
        - apiGroups:
          - '*'
          resources:
          - '*'
          verbs:
          - '*'
        serviceAccountName: consul-operator
    strategy: deployment
  installModes:
  - supported: true
    type: OwnNamespace
  - supported: true
    type: SingleNamespace
  - supported: false
    type: MultiNamespace
  - supported: true
    type: AllNamespaces
  maturity: alpha
  provider: {}
  version: 0.0.10
```
The deployment created by the olm-operator based on the CSV, with empty resources section:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "1"
  creationTimestamp: "2020-12-17T09:15:42Z"
  generation: 4
  labels:
    olm.deployment-spec-hash: 65fb759884
    olm.owner: consul.v0.0.10
    olm.owner.kind: ClusterServiceVersion
    olm.owner.namespace: consul-stable
    operators.coreos.com/consul.consul-stable: ""
  name: consul-operator
  namespace: consul-stable
  ownerReferences:
  - apiVersion: operators.coreos.com/v1alpha1
    blockOwnerDeletion: false
    controller: false
    kind: ClusterServiceVersion
    name: consul.v0.0.10
    uid: edc91de1-e2c1-46c1-b5b3-613c717e7e06
  resourceVersion: "46735524"
  selfLink: /apis/apps/v1/namespaces/consul-stable/deployments/consul-operator
  uid: 02e0db72-4933-4303-9b53-661f8438bc14
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 1
  selector:
    matchLabels:
      name: consul-operator
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        capabilities: Basic Install
        olm.operatorGroup: consul
        olm.operatorNamespace: consul-stable
        olm.targetNamespaces: consul-stable
        operatorframework.io/properties: '{"properties":[{"type":"olm.gvk","value":{"group":"application.com","kind":"Consul","version":"v1alpha1"}},{"type":"olm.gvk","value":{"g                              roup":"application.com","kind":"Consul","version":"v1alpha2"}},{"type":"olm.package","value":{"packageName":"consul","version":"0.0.10"}}]}'
      creationTimestamp: null
      labels:
        name: consul-operator
    spec:
      containers:
      - command:
        - consul-operator
        env:
        - name: WATCH_NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.annotations['olm.targetNamespaces']
        - name: POD_NAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.name
        image: consul-operator:1.19
        imagePullPolicy: IfNotPresent
        name: consul-operator
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      imagePullSecrets:
      - name: pull-secret
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: consul-operator
      serviceAccountName: consul-operator
      terminationGracePeriodSeconds: 30
status:
  conditions:
  - lastTransitionTime: "2020-12-17T09:15:42Z"
    lastUpdateTime: "2020-12-17T09:15:42Z"
    message: Deployment does not have minimum availability.
    reason: MinimumReplicasUnavailable
    status: "False"
    type: Available
  - lastTransitionTime: "2020-12-17T09:15:42Z"
    lastUpdateTime: "2020-12-17T09:15:42Z"
    message: 'pods "consul-operator-88f8cb9b-8mbvx" is forbidden: failed quota:
      consul-stable: must specify cpu,memory'
    reason: FailedCreate
    status: "True"
    type: ReplicaFailure
  - lastTransitionTime: "2020-12-17T09:25:42Z"
    lastUpdateTime: "2020-12-17T09:25:42Z"
    message: ReplicaSet "consul-operator-88f8cb9b" has timed out progressing.
    reason: ProgressDeadlineExceeded
    status: "False"
    type: Progressing
  observedGeneration: 4
    unavailableReplicas: 1
```

**What did you expect to see?**
If there is no resources config defined in the subscription then the resources defined in the CSV should be preserved.

**What did you see instead? Under which circumstances?**
The resources section in the deployment was overwritten with the empty section from the subscription.

**Environment**
* operator-lifecycle-manager version: 0.17.0

* Kubernetes version information:

vagrant@vmvagrant> kubectl version
Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.2", GitCommit:"66049e3b21efe110454d67df4fa62b08ea79a19b", GitTreeState:"clean", BuildDate:"2019-05-16T16:23:09Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.9", GitCommit:"a17149e1a189050796ced469dbd78d380f2ed5ef", GitTreeState:"clean", BuildDate:"2020-04-16T11:36:15Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}

* Kubernetes cluster kind: 

**Possible Solution**

It is not checked if the resource config was defined in the subscription and the resources section of the containers is overwritten with the empty structure.

**Additional context**
https://github.com/operator-framework/operator-lifecycle-manager/pull/1334#issuecomment-747382274 

@umangachapagain

### LINKED ISSUE #1940: Operator deployment resources claimed in CSV aren't passed into the operator deployment
## Bug Report

**What did you do?**
A clear and concise description of the steps you took (or insert a code snippet).

I try to deploy an Operator into the Openshift cluster. After the operator is deployed, I found the `.spec.containers[].resources` claimed in the `csv.spec.install.deployments` are not added in the operator deployment.
 
For example,
CSV:
```
  install:
    spec:
      deployments:
       ...
                resources:
                  limits:
                    cpu: 200m
                    memory: 256Mi
                  requests:
                    cpu: 200m
                    memory: 256Mi
```
Deployment:
```
    spec:
    ...
      containers:
       ...
        - resources: {}
```

**What did you expect to see?**
A clear and concise description of what you expected to happen (or insert a code snippet).

I expect the deployment `resources` added into the CSV can be added into operator deployment during the installation.
 
**What did you see instead? Under which circumstances?**
A clear and concise description of what you expected to happen (or insert a code snippet).

The resources field is empty in the operator deployment. 

**Environment**
* operator-lifecycle-manager version:
0.15.1

* Kubernetes version information:
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.8", GitCommit:"9f2892aab98fe339f3bd70e3c470144299398ace", GitTreeState:"clean", BuildDate:"2020-08-13T16:12:48Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.3+ca4017d", GitCommit:"ca4017d", GitTreeState:"clean", BuildDate:"2020-11-14T01:37:24Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}

* Kubernetes cluster kind: 
OCP
**Possible Solution**

**Additional context**

### LINKED ISSUE #1953: Subscription with empty resource config will overwirte operator deployment resource
## Bug Report

**What did you do?**
Just install the olm

**What did you expect to see?**
the builtint packageserver operator with resources settings in it's csv and i hoping to see it in the actual deployment
image

**What did you see instead? Under which circumstances?**
packageserver deployment with empty resoruce
image

**Environment**
* operator-lifecycle-manager version:  

0.17.0

* Kubernetes version information:

v1.16.9

* Kubernetes cluster kind: 

**Possible Solution**

**Additional context**
I installl some of my own operators, same thing happend

### PR DESCRIPTION
**Pull request by joelanford**
**Description of the change:**
This PR resolves an issue that causes resources defined in CSV deployment specs to be removed by OLM when the operator is installed unless they are explicitly redefined in the subscription's `spec.config.resources` section.

**Motivation for the change:**
Closes #1912  
Closes #1940 
Closes #1953 

**Reviewer Checklist**
- [x] Implementation matches the proposed design, or proposal is updated to match implementation
- [x] Sufficient unit test coverage 
- [x] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [x] Commit messages sensible and descriptive

### CODE REVIEW

**Comment by joelanford** [comment]
Note: This PR is current a WIP because it needs some (hopefully quick) discussion about the approach.

There are likely two ways to solve this problem:
1. Make `subscription.spec.config.resources` a pointer and do a `nil` check before applying the override. That's what this PR currently does, however it is a breaking change to the Go API (not to the kubernetes API)
2. Leave `subscription.spec.config.resources` as it is, and do `reflect.DeepEqual(overrides, corev1.ResourceRequirements{})` to perform the comparison instead of the `nil` check. This option has the benefit of not changing the Go API code at all. However it feels hacky and workaround-y.

My vote is for #1 unless there's a major concern with changing the Go API.

Also note that this PR goes ahead and changes the entire `subscription.spec.config` struct to a pointer as well since `config` is optional.

**Comment by joelanford** [comment]
I think I finally figured out what's going on here. operator-framework/api@v0.5.2 requires controller-runtime v0.8.0 (olm master is currently on v0.5.1). Trying to apply a necessary change to that repo and pull it into this PR means we're also getting v0.5.3+ and controller-runtime v0.8.0.

The problem is that v0.8.0 defines an interface that contains overlapping interfaces, which is only supported in Go 1.14+. The OLM `go.mod` file still specifies Go 1.13, so even if we're compiling with a later version, we are still getting Go 1.13 semantics.

It seems like fixing this will require a Go version bump for OLM, which is probably overdue anyway.

Just as a note to our future selves, we should bump the minor version on operator-framework/api releases when we change Go, Kubernetes, or controller-runtime minor versions (and any others we think might be troublesome), so that we don't get into a situation (like this one) where we need to fix a bug in OLM that requires an API change, but now all of a sudden we're pulling in a bunch of other unrelated changes.
/cc  @estroz, @dinhxuanvu

**Comment by estroz** [comment]
> Just as a note to our future selves, we should bump the minor version on operator-framework/api releases when we change Go, Kubernetes, or controller-runtime minor versions

Fully agree and have brought this up before offline. I cut v0.5 as the “k8s 1.20” release although v0.5.0 still used controller-runtime v0.7.0 because v0.8.0 wasn’t out yet, which I think is generally fine but led to issues due to min module versioning.

**Comment by joelanford** [comment]
This PR needs to be rebased after https://github.com/operator-framework/operator-lifecycle-manager/pull/2012 merges.

File: go.mod
Comment (njhale) [inline go.mod:67]: note: just need to remove this and bump the `operator-framework/api` version to a new release including the API change.

File: pkg/controller/operators/olm/overrides/config.go
Comment (njhale) [inline pkg/controller/operators/olm/overrides/config.go:39]: nit: less indentation with something like

```go
if owner.Spec.Config == nil {
    // No overrides
    return
}

envVarOverrides = owner.Spec.Config.Env
volumeOverrides = owner.Spec.Config.Volumes
volumeMountOverrides = owner.Spec.Config.VolumeMounts
tolerationOverrides = owner.Spec.Config.Tolerations
resourcesOverride = owner.Spec.Config.Resources
nodeSelectorOverride = owner.Spec.Config.NodeSelector
```

Also, we probably want to check this up front to avoid unnecessary list/filter operations.

Comment (joelanford) [reply pkg/controller/operators/olm/overrides/config.go:39]: +1 on reducing indentation.

> Also, we probably want to check this up front to avoid unnecessary list/filter operations.

In this function, we list the subscriptions, then find the subscription that owns the CSV. `owner` is this subscription, so the `owner.Spec.Config` check requires the list and filter to happen if I'm understanding this correctly. Is there something further up the call stack you're referring to?

**Review by dinhxuanvu — APPROVED** [review]
/lgtm

**Comment by njhale** [comment]
/hold

console tests consistently failing -- waiting for [more info from the test maintainers](https://coreos.slack.com/archives/C6A3NV5J9/p1614696443124200)

**Comment by taylormgeorge91** [comment]
Will this be targeted for release-4.6 backport or only 4.7?

**Comment by horis233** [comment]
@taylormgeorge91 
This is the cherry-pick PR for release-4.6 https://github.com/operator-framework/operator-lifecycle-manager/pull/2071

**Pull request merged**
